tests/tests.cpp

```diff
@@ -0,0 +1,398 @@
+        ASSERT_EQ(0, crypto.SymEnc(sym_key, msg.data(),
+                                   reinterpret_cast<const uint8_t*>("aad"),
+                                   ct.data(), msg.size(), strlen("aad")));
+
+        // We don't care about the value of the decryptions (they all should
+        // fail) so we can use the same plaintext vector for each decryption
+        std::vector<uint8_t> pt(crypto.SymDecSize(ct.size()));
+
+        // 1) Make random pertubations in the ciphertexts
+        for (int i = 0; i < 100; i++) {
+            std::vector<uint8_t> perturbed_ct(ct);
+            perturb(perturbed_ct);
+            ASSERT_NE(0, crypto.SymDec(sym_key, perturbed_ct.data(),
+                                       reinterpret_cast<const uint8_t*>("aad"),
+                                       pt.data(), perturbed_ct.size(),
+                                       strlen("aad")))
+                << "Perturbed ciphertext decrypted successfully";
+        }
+
+        // 2) Change the IV
+        uint8_t new_iv[CIPHER_IV_SIZE];
+        crypto.RandGen(new_iv, CIPHER_IV_SIZE);
+
+        std::vector<uint8_t> perturbed_ct(ct);
+        for (int i = 0; i < CIPHER_IV_SIZE; i++) {
+            perturbed_ct[i] = new_iv[i];
+        }
+        ASSERT_NE(0,
+                  crypto.SymDec(sym_key, perturbed_ct.data(),
+                                reinterpret_cast<const uint8_t*>("aad"),
+                                pt.data(), perturbed_ct.size(), strlen("aad")))
+            << "Ciphertext with invalid IV decrypted successfully";
+
+        // 3) Use incorrect AAD
+        ASSERT_NE(0, crypto.SymDec(sym_key, ct.data(),
+                                   reinterpret_cast<const uint8_t*>("aa"),
+                                   pt.data(), ct.size(), strlen("aa")))
+            << "Ciphertext with invalid AAD decrypted successfully";
+
+        // 4) Decrypt with the wrong key
+        uint8_t new_key[CIPHER_KEY_SIZE];
+        crypto.RandGen(new_key, CIPHER_KEY_SIZE);
+
+        ASSERT_NE(0, crypto.SymDec(new_key, ct.data(),
+                                   reinterpret_cast<const uint8_t*>("aad"),
+                                   pt.data(), ct.size(), strlen("aad")))
+            << "Ciphertext decrypted successfully with incorrect key";
+
+        // 5) Remove last block of ciphertext
+        perturbed_ct = std::vector<uint8_t>(ct);
+        perturbed_ct.resize(perturbed_ct.size() - CIPHER_KEY_SIZE);
+
+        ASSERT_NE(0,
+                  crypto.SymDec(sym_key, perturbed_ct.data(),
+                                reinterpret_cast<const uint8_t*>("aad"),
+                                pt.data(), perturbed_ct.size(), strlen("aad")))
+            << "Truncated ciphertext decrypted successfully";
+
+        // 6) Swap random blocks in the ciphertext
+        for (int i = 0; i < 100; i++) {
+            perturbed_ct = std::vector<uint8_t>(ct);
+            std::vector<uint8_t> ct_blocks(perturbed_ct.begin() +
```

Comment:
can you add some clarifying comments on what you're doing here? e.g. `we swap blocks only in `ct_block`, which is the portion of the ciphertext that doesn't contain the IV/TAG

Reply:
Good idea - added